### PR TITLE
Entropy Ack

### LINKED ISSUE #127: Handle out-of band EntropyRequest
**Describe the bug**
Handle out-of-band entropy requests.

**Environment information:**

- OS: all
- Platform: discovered on `Linux name 4.19.0-1-amd64 1 SMP Debian 4.19.12-1 (2018-12-22) x86_64 GNU/Linux`
- Firmware Version: skycoin/hardware-wallet@ffd6fcd 

**Steps to Reproduce**

Run integration tests against physical Skywallet device .

**Actual behavior**

This is the output of a `go run cmd/cli/cli.go features` for example:

```
[2019-05-10T21:28:35-04:00] INFO [walletusb]: bLength 18
bDescriptorType 1
bcdUSB 2.00
bDeviceClass 0
bDeviceSubClass 0
bDeviceProtocol 0
bMaxPacketSize0 64
idVendor 0x313a
idProduct 0x0001
bcdDevice 1.00
iManufacturer 1
iProduct 2
iSerialNumber 3
bNumConfigurations 1
[2019-05-10T21:28:35-04:00] ERROR [walletusb]: current configuration 1
[2019-05-10T21:28:35-04:00] INFO [walletusb]: finishing read queue
[2019-05-10T21:28:35-04:00] ERROR [skycoin-hw-cli]: received unexpected message type: MessageType_EntropyRequest
```

**Expected behavior**
No matter what command is executing , the entropy message should be either ignored (not recommended) or handled by sending `EntropyAck` back to the physical device.

**Additional context**
Initially reported in https://github.com/skycoin/hardware-wallet-go/pull/126#issuecomment-491469255

**Possible implementation**
beware of the fact that this affects all possible interactions with the device , so maybe 

- [ ] Handle `EntropyRequest` gracefully by ignoring it
- [ ] If handling it, add a separate goroutine to receive `EntropyRequest` and reply with `EntropyAck`

### PR DESCRIPTION
**Pull request by therealssj**
Fixes #127 

Changes:
- Handle entropy ack request

Does this change need to mentioned in CHANGELOG.md?
no

Requires changes in protobuff specs?
no

Requires testing
yes

### CODE REVIEW

File: src/skywallet/skywallet.go [hunk 1]
```diff
@@ -185,12 +182,32 @@ func (d *Device) AddressGen(addressN, startIndex uint32, confirmAddress bool) (w
...
 		return wire.Message{}, err
 	}
 
-	return d.Driver.SendToDevice(d.dev, chunks)
+	msg, err := d.Driver.SendToDevice(d.dev, addressGenChunks)
+	if err != nil {
+		return msg, err
+	}
+
+	if msg.Kind == uint16(messages.MessageType_MessageType_EntropyRequest) {
```
Comment (olemis) [inline src/skywallet/skywallet.go]: can this be implemented in d.Driver.SendToDevice ?

Comment (therealssj) [reply src/skywallet/skywallet.go]: I have added a generic function now which will send msg to the device and handle the response
https://github.com/skycoin/hardware-wallet-go/pull/129/files#diff-35c9248b784427ee5ad156f6636bf91cR805

File: src/skywallet/helper.go [hunk 2]
```diff
@@ -187,17 +187,42 @@ func sendToDeviceNoAnswer(dev usb.Device, chunks [][64]byte) error {
...
+
+			for _, element := range entropyChunks {
+				_, err := dev.Write(element[:])
+				if err != nil {
+					log.Errorf("entropy ack error: %v", err)
+				}
+			}
+		}()
+
+		msg, err = wire.ReadFrom(dev)
```
Comment (therealssj) [inline src/skywallet/helper.go]: @olemis @stdevAlDen is it possible to read again or do I need to send the old chunks again
```
for _, element := range chunks {
			_, err := dev.Write(element[:])
			if err != nil {
				return wire.Message{}, err
			}
		}
```
and then read it.

I am not sure if this works with the goroutine writing in parallel.
@stdevAlDen can you test this for me?

Comment (olemis) [reply src/skywallet/helper.go]: Read again . `EntropyRequest` is sent out-of-band while other messages might be undergoing processing or waiting for user input.

Comment (olemis) [reply src/skywallet/helper.go]: if you send chunks again then firmware interprets you are retrying the same command once again

Comment (stdevAlDen) [reply src/skywallet/helper.go]: > @olemis @stdevAlDen is it possible to read again or do I need to send the old chunks again
> 
> ```go
> for _, element := range chunks {
> 	_, err := dev.Write(element[:])
> 	if err != nil {
> 		return wire.Message{}, err
> 	}
> }
> ```
> 
> and then read it.
> 
> I am not sure if this works with the goroutine writing in parallel.
> @stdevAlDen can you test this for me?

@therealssj 
It's a bit difficult to make collide a a user command with a call to get external entropy at the same instant, to force this I use a long for loop to force this happen:
`for i in $(seq 300000) ; do skycoin-hw-cli features > /tmp/a ; done`

From time to time this happened, I can not find any relevant info the know if this is related 
 to the original issue, please check out the log:
```bash
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x6e30ef]

runtime stack:
runtime.throw(0x798a4d, 0x2a)
        /usr/lib/go/src/runtime/panic.go:617 +0x72
runtime.sigpanic()
        /usr/lib/go/src/runtime/signal_unix.go:374 +0x4a9

goroutine 6 [syscall]:
runtime.cgocall(0x6ef2a0, 0xc000158d88, 0x0)
        /usr/lib/go/src/runtime/cgocall.go:128 +0x5b fp=0xc000158d58 sp=0xc000158d20 pc=0x40869b
github.com/skycoin/hardware-wallet-go/src/usb/lowlevel/libusb._Cfunc_libusb_interrupt_transfer(0x1c46760, 0x1c46701, 0xc0001521c0, 0x40, 0xc00011ba74, 0x0, 0xc000000000)
        _cgo_gotypes.go:1180 +0x4d fp=0xc000158d88 sp=0xc000158d58 pc=0x67a5ed
github.com/skycoin/hardware-wallet-go/src/usb/lowlevel/libusb.Interrupt_Transfer.func1(0x1c46760, 0xc00011ba01, 0xc0001521c0, 0x40, 0x40, 0xc00011ba74, 0x0, 0x3)
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/src/usb/lowlevel/libusb/libusb.go:1427 +0x93 fp=0xc000158dd0 sp=0xc000158d88 pc=0x6808c3
github.com/skycoin/hardware-wallet-go/src/usb/lowlevel/libusb.Interrupt_Transfer(0x1c46760, 0xdba01, 0xc0001521c0, 0x40, 0x40, 0x0, 0x10000c000158ef8, 0x0, 0x40, 0xc0001521c0, ...)
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/src/usb/lowlevel/libusb/libusb.go:1427 +0x85 fp=0xc000158e30 sp=0xc000158dd0 pc=0x67fa85
github.com/skycoin/hardware-wallet-go/src/skywallet/usb.(*LibUSBDevice).readWrite(0xc00000ee80, 0xc0001521c0, 0x40, 0x40, 0x1, 0x725de0, 0x1, 0xc0001521c0)
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/src/skywallet/usb/libusb.go:439 +0x9b fp=0xc000158eb8 sp=0xc000158e30 pc=0x683b7b
github.com/skycoin/hardware-wallet-go/src/skywallet/usb.(*LibUSBDevice).Write(0xc00000ee80, 0xc0001521c0, 0x40, 0x40, 0x0, 0x0, 0x0)
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/src/skywallet/usb/libusb.go:475 +0x61 fp=0xc000158f08 sp=0xc000158eb8 pc=0x683f01
github.com/skycoin/hardware-wallet-go/src/skywallet.sendToDevice.func1(0x7f55c0, 0xc00000ee80)
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/src/skywallet/helper.go:213 +0x138 fp=0xc000158fd0 sp=0xc000158f08 pc=0x6d34f8
runtime.goexit()
        /usr/lib/go/src/runtime/asm_amd64.s:1337 +0x1 fp=0xc000158fd8 sp=0xc000158fd0 pc=0x45c801
created by github.com/skycoin/hardware-wallet-go/src/skywallet.sendToDevice
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/src/skywallet/helper.go:205 +0x211

goroutine 1 [runnable]:
fmt.(*pp).handleMethods(0xc0000a6000, 0x20300000000073, 0x7f40cc412f00)
        /usr/lib/go/src/fmt/print.go:606 +0x100
fmt.(*pp).printArg(0xc0000a6000, 0x73f480, 0xc000168018, 0xc000000073)
        /usr/lib/go/src/fmt/print.go:699 +0x20a
fmt.(*pp).doPrintf(0xc0000a6000, 0x79694b, 0x24, 0xc00016a000, 0x1, 0x1)
        /usr/lib/go/src/fmt/print.go:1016 +0x158
fmt.Sprintf(0x79694b, 0x24, 0xc00016a000, 0x1, 0x1, 0xb43660, 0x7f40ce424d98)
        /usr/lib/go/src/fmt/print.go:214 +0x66
github.com/skycoin/hardware-wallet-go/vendor/github.com/sirupsen/logrus.(*Entry).Logf(0xc0000c01c0, 0xc000000002, 0x79694b, 0x24, 0xc00016a000, 0x1, 0x1)
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/vendor/github.com/sirupsen/logrus/entry.go:314 +0x7c
github.com/skycoin/hardware-wallet-go/vendor/github.com/sirupsen/logrus.(*Entry).Errorf(0xc0000c01c0, 0x79694b, 0x24, 0xc00016a000, 0x1, 0x1)
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/vendor/github.com/sirupsen/logrus/entry.go:343 +0x65
github.com/skycoin/hardware-wallet-go/src/cli.featuresCmd.func1(0xc00009e420)
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/src/cli/features.go:72 +0x5fa
github.com/skycoin/hardware-wallet-go/vendor/github.com/urfave/cli.HandleAction(0x71c860, 0x79fa68, 0xc00009e420, 0xc00006e400, 0x0)
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/vendor/github.com/urfave/cli/app.go:492 +0x7c
github.com/skycoin/hardware-wallet-go/vendor/github.com/urfave/cli.Command.Run(0x78cb91, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x791b2c, 0x18, 0x0, ...)
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/vendor/github.com/urfave/cli/command.go:210 +0x996
github.com/skycoin/hardware-wallet-go/vendor/github.com/urfave/cli.(*App).Run(0xc00009ab60, 0xc00000e080, 0x2, 0x2, 0x0, 0x0)
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/vendor/github.com/urfave/cli/app.go:255 +0x6af
github.com/skycoin/hardware-wallet-go/src/cli.(*App).Run(...)
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/src/cli/cli.go:123
main.main()
        /usr/share/gocode/src/github.com/skycoin/hardware-wallet-go/cmd/cli/cli.go:17 +0xd1
```

**Comment by therealssj** [comment]
@stdevAlDen can you try now?

**Comment by stdevAlDen** [comment]
> @stdevAlDen can you try now?

@therealssj seems as working right :smile:

**Pull request merged**